File: ScoreFrame-Android/app/src/main/java/com/scoreframe/model/TimeoutEvent.kt
package com.scoreframe.model

import kotlinx.serialization.Serializable
import kotlin.math.max

@Serializable
data class TimeoutEvent(
    val id: String = java.util.UUID.randomUUID().toString(),
    val team: Team,
    val timestamp: Double,
    var endTimestamp: Double? = null
) {
    fun isActive(videoTime: Double): Boolean {
        if (videoTime < timestamp) return false
        val end = endTimestamp ?: return true
        return videoTime < end
    }

    fun elapsedSeconds(videoTime: Double): Double? {
        if (videoTime < timestamp) return null
        val end = endTimestamp
        if (end != null && videoTime >= end) return null
        return videoTime - timestamp
    }

    fun pausedSeconds(kickoff: Double, videoTime: Double): Double {
        val overlapStart = max(timestamp, kickoff)
        val overlapEnd = if (endTimestamp != null) {
            minOf(endTimestamp!!, videoTime)
        } else {
            videoTime
        }
        return max(0.0, overlapEnd - overlapStart)
    }
}

File: OverScore/OverScore/Models/TimeoutEvent.swift
import Foundation

struct TimeoutEvent: Codable, Identifiable, Equatable {
    var id: UUID = UUID()
    var team: Team
    var timestamp: TimeInterval       // タイムアウト開始（動画内時刻）
    var endTimestamp: TimeInterval?    // タイムアウト終了（nil = まだ進行中）

    func isActive(at videoTime: TimeInterval) -> Bool {
        guard videoTime >= timestamp else { return false }
        if let end = endTimestamp { return videoTime < end }
        return true // 終了未設定 = 進行中
    }

    func elapsedSeconds(at videoTime: TimeInterval) -> TimeInterval? {
        guard videoTime >= timestamp else { return nil }
        if let end = endTimestamp, videoTime >= end { return nil }
        return videoTime - timestamp
    }

    /// kickoff〜videoTime の間にこのタイムアウトが停止させた累計秒数
    func pausedSeconds(from kickoff: TimeInterval, to videoTime: TimeInterval) -> TimeInterval {
        let overlapStart = max(timestamp, kickoff)
        let overlapEnd: TimeInterval
        if let end = endTimestamp {
            overlapEnd = min(end, videoTime)
        } else {
            overlapEnd = videoTime
        }
        return max(0, overlapEnd - overlapStart)
    }
}
